mpc-core/src/protocols/rep3_ring/gadgets/sort.rs

```diff
@@ -0,0 +1,625 @@
+    apply_inv_field(&perm, inputs, io_context)
+}
+
+fn gen_perm<F: PrimeField, N: Rep3Network>(
+    inputs: &[FieldShare<F>],
+    io_context: &mut IoContext<N>,
+    bitsize: usize,
+) -> IoResult<Vec<Rep3RingShare<PermRing>>> {
+    let mask = (BigUint::one() << bitsize) - BigUint::one();
+
+    // Decompose
+    // TODO: This step could be optimized (I: Pack the a2b's, II: only reconstruct bitsize bits)
+    let mut bits = Vec::with_capacity(bitsize);
+    for inp in inputs.iter().cloned() {
+        let mut binary = rep3::conversion::a2b_selector(inp, io_context)?;
+        binary &= &mask;
+        bits.push(binary)
+    }
+
+    let bit_0 = inject_bit(&bits, io_context, 0)?;
+    let mut perm = gen_bit_perm(bit_0, io_context)?;
+
+    for i in 1..bitsize {
+        let bit_i = inject_bit(&bits, io_context, i)?;
+        let bit_i = apply_inv(&perm, &bit_i, io_context)?;
+        let perm_i = gen_bit_perm(bit_i, io_context)?;
+        perm = compose(perm, perm_i, io_context)?;
+    }
+
+    Ok(perm)
+}
+
+fn inject_bit<F: PrimeField, N: Rep3Network>(
+    inputs: &[Rep3BigUintShare<F>],
+    io_context: &mut IoContext<N>,
+    bit: usize,
+) -> IoResult<Vec<Rep3RingShare<PermRing>>> {
+    let len = inputs.len();
+    let mut bits = Vec::with_capacity(len);
+    for inp in inputs {
+        let a = inp.a.bit(bit as u64) as PermRing;
+        let b = inp.b.bit(bit as u64) as PermRing;
+        bits.push(Rep3RingShare::new_ring(a.into(), b.into()));
+    }
+    conversion::bit_inject_many(&bits, io_context)
+}
+
+fn gen_bit_perm<N: Rep3Network>(
+    bits: Vec<Rep3RingShare<PermRing>>,
+    io_context: &mut IoContext<N>,
+) -> IoResult<Vec<Rep3RingShare<PermRing>>> {
+    let len = bits.len();
+    let mut f0 = Vec::with_capacity(len);
+    let mut f1 = Vec::with_capacity(len);
+    for inp in bits {
+        f0.push(arithmetic::add_public(
+            -inp,
+            RingElement::one(),
+            io_context.id,
+        ));
+        f1.push(inp);
+    }
+
+    let mut s = Rep3RingShare::zero_share();
+    let mut s0 = Vec::with_capacity(len);
+    let mut s1 = Vec::with_capacity(len);
+    for f in f0.iter() {
+        s = arithmetic::add(s, *f);
+        s0.push(s);
+    }
+    for f in f1.iter() {
+        s = arithmetic::add(s, *f);
+        s1.push(s);
+    }
+    let mul1 = arithmetic::mul_vec(&f0, &s0, io_context)?;
+    let mul2 = arithmetic::mul_vec(&f1, &s1, io_context)?;
+    let perm = mul1
+        .into_iter()
+        .zip(mul2)
+        .map(|(a, b)| arithmetic::add(a, b))
+        .collect();
+
+    Ok(perm)
+}
+
+fn apply_inv<T: IntRing2k, N: Rep3Network>(
+    rho: &[Rep3RingShare<PermRing>],
+    bits: &[Rep3RingShare<T>],
+    io_context: &mut IoContext<N>,
+) -> IoResult<Vec<Rep3RingShare<T>>>
+where
+    Standard: Distribution<T>,
+{
+    let len = rho.len();
+    debug_assert_eq!(len, rho.len());
+
+    let unshuffled = (0..len as PermRing).collect::<Vec<_>>();
```

Comment:
Can we only perm 4GB?

Reply:
have added a check in the beginning. We get an error when the size is too large and an instruction to increase ring size. Doubling ring size unfortuantely also doubles the communication, so I wanted to keep it restricted. I think sorting 4*10^9 elements is enough for now (in shared form the have 32 * 4* 10^9 * 2 bytes, i.e., 256 GB of data, so RAM will be high anyways)

Reply:
I added a comment to the Perm ring as well